### DISCUSSION TITLE
关于自建服务器的问题

### DISCUSSION THREAD

**Discussion by xiebruce** [chanify/chanify#133] [Q&A]
- 1、作为无状态服务器的意义在哪儿？自建服务器我觉得一是为了安全，二是减少开发者的服务器太多人用承受不住，既然都要通过api.chanify.net，那跟我直接客户端用api.chanify.net有什么区别呢？
- 2、配置文件中，endpoint有什么用？如果说用http，那么我直接用ip也好，用域名也好，访问就行了，如果是https，我用nginx转发到127.0.0.1:8080中就行，搞不懂这个endpoint有什么用，目前发现的唯一用处就是反作用：增加配置难度，外网访问的链接必须填到endpoint中，否则扫码就会显示失效；
- 3、[configuration](https://github.com/chanify/chanify#configuration)配置文件，不作为客户端使用时，client部分的配置是不是可以不写？client部分的具体作用是什么？是相当于Chanify app的命令行版本吗？
- 4、[configuration](https://github.com/chanify/chanify#configuration)配置文件中哪些参数不是必须的？比如`dburl`，是必须的吗？我感觉不是吧？
- 5、[configuration](https://github.com/chanify/chanify#configuration)配置文件中`datapath`和`dburl`是不是二者取其一就行？
- 6、[configuration](https://github.com/chanify/chanify#configuration)配置文件中register的enable设置为false，并且不设置白名单，用Chanify扫码，一样可以添加，这算是register吗？如果算，那么这个enable设置为false根本没用，如果不算，那什么才是register？

文档我看了，感觉好多都没有说明，又或者是这个工具太复杂功能太多，我理解不了？

**Comment by wizjin** [comment]
1. 首先根据Chanify的加密逻辑，即使数据通过api.chanify.net，在api.chanify.net也是无法解密数据的。所以无状态服务器的意义在于独立的加密密钥，保证安全。有状态的用于可以发送文件图片。
2. endpoint的意义在于定义外部访问地址，分发的消息里面如果包含发送文件地址时候用于下载图片的。
3. 不作为客户端使用时，client部分的配置可以不用写。
4. dburl不是必须，没有配置会自动生成sqlite的数据
5. datapath和dburl用处不一样，datapath用于保存上传图片的目录，dburl用于连接数据库
6. register只有在有状态的服务器模式才会起作用，在有状态模式下register的enable设置为false时，用Chanify扫码就不能添加。

**Reply by xiebruce** [reply]
感谢作者大佬的回答，原来图片、文件是用链接的形式的，还会保存在服务器，这个我没配置到，确实不知道，1、2、4、5已经了解了，但是3、6还是有问题，先说6，我这边肯定是有状态的吧？因为我command加了`serve`
<img width="508" alt="image" src="https://user-images.githubusercontent.com/13575500/201683428-a1bd7bc7-d8e7-45c9-8e53-c51abc57f9f9.png">

这是我的配置文件，我又重启了一遍chanify，app删掉了之前添加的重新添加的，真的能添加上，理论上来说不应该能添加上的吧？
<img width="1034" alt="image" src="https://user-images.githubusercontent.com/13575500/201685799-5c7af5ef-4789-4ddd-b081-5fab370b47f2.png">

`docker ps -a --no-trunc`查看，可以证明确实是加了`serve`
<img width="1288" alt="image" src="https://user-images.githubusercontent.com/13575500/201687241-ef045242-11d4-49ed-9c79-aca2a2a5165c.png">

---

而3的话，我还是没懂这“客户端”的意思，我所理解的客户端，比如我现在配置的这个server模块，属于服务器端，而对应的客户端是Chanify app，所以如果在这里配置`client`模块，我的理解，它就是“命令行版”客户端了，从你这文档看也是这样的
<img width="1080" alt="image" src="https://user-images.githubusercontent.com/13575500/201689291-720a2c3a-a48b-4504-ac74-2ea92bc4da92.png">

但是，你看[这篇文章](https://post.smzdm.com/p/akx9l0nr/)，你页内搜索：`配置PC端，实现PC上推送文件到手机`，它在PC端添加之后，又在群晖上添加`client`配置，这时候我的理解是，它已经没什么“客户端”的概念了，而是两台机之间可以相互推送文件和消息，跟server模块没啥关系了，是不？

**Comment by wizjin** [comment]
是不是有状态的可以看一下扫描二维码之后的状态列表，有存储设备信息才表示是有状态的。
有状态和是不是加了serve没关系，也可以看一下日志，里面有显示的。

这个client是表示用于发送消息的客户端，App是接收消息的客户端。

**Comment by xiebruce** [comment]
我扫码后，那个存储设备信息按钮是有的，但默认是没有开的，是需要我手动开吗？

我对比你文档中无状态和有状态的两个启动命令，发现区别是，有状态的不用secert，但是有一个datapath，而无状态有secret无datapath
```bash
chanify serve --host=<ip address> --port=<port> --secret=<secret key> --name=<node name> --endpoint=http://<address>:<port>
chanify serve --host=<ip address> --port=<port> --name=<node name> --datapath=~/.chanify --endpoint=http://<address>:<port>
```

我启动的时候是有指定datapath的，看了日志也是serverless:false，双重否定那就是有状态了，我确认了一下，register enable为false且不设置白名单时，真的能添加上，你看这日志
```bash
2022/11/16 18:04:10 Shutting down server...
2022/11/16 18:04:10 Shutdown service success.
2022/11/16 18:04:10 Launching service...
2022/11/16 18:04:10 Using config file: /root/.chanify.yml
2022/11/16 18:04:10 Register user disabled
2022/11/16 18:04:10 Find 0 user(s) in whitelist
2022/11/16 18:04:10 Open sqlite database: /data/chanify.db
2022/11/16 18:04:10 Files path: /data/files
2022/11/16 18:04:10 Load plugins: /root/.chanify/plugin
2022/11/16 18:04:10 Node server name: Bruce's Chanify Server, version: 1.4.2, serverless: false, node-id: AEMIEYPWDDXWGJVMSJOTIUUDBFECUDJPKA
2022/11/16 18:04:10 Launch service 127.0.0.1:8081
2022/11/16 18:04:10 Node server endpoint: https://pnz2-chanify.xiebruce.cf
2022/11/16 18:04:29 200 |  113.91.210.211 | GET /rest/v1/info   69.382µs "Chanify/1.4.5-68 (iPhone12,1; iOS 16.1; Scale/2.00)"
2022/11/16 18:04:32 Bind user: ABLGYLOWPHIQFWHY6XZXEVTMGWQ44YZ7VM
2022/11/16 18:04:32 200 |  113.91.210.211 | POST /rest/v1/bind-user 1.359715ms "Chanify/1.4.5-68 (iPhone12,1; iOS 16.1; Scale/2.00)"
```

---

另外，client我能理解了，谢谢

**Reply by wizjin** [reply]
应该是之前在register enable为true的时候这个手机已经添加过了，添加过的手机不会受register enable影响。
可以删除chanify.db，应该就不会添加成功。

**Reply by xiebruce** [reply]
> 应该是之前在register enable为true的时候这个手机已经添加过了，添加过的手机不会受register enable影响。 可以删除chanify.db，应该就不会添加成功。

原来如此，这样就说的通了。
